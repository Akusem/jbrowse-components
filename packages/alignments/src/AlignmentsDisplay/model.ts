import {
  getConf,
  ConfigurationReference,
} from '@gmod/jbrowse-core/configuration'
import { BaseTrack } from '@gmod/jbrowse-plugin-linear-genome-view'
import { types, addDisposer } from 'mobx-state-tree'
import { autorun } from 'mobx'
import AlignmentsDisplayComponent from './components/AlignmentsDisplay'

// eslint-disable-next-line @typescript-eslint/no-explicit-any
export default (pluginManager: any, configSchema: any) => {
  return types
    .compose(
      'AlignmentsDisplay',
      BaseTrack,
      types
        .model({
          PileupDisplay: types.maybe(
            pluginManager.getDisplayType('PileupDisplay').stateModel,
          ),
          SNPCoverageDisplay: types.maybe(
            pluginManager.getDisplayType('SNPCoverageDisplay').stateModel,
          ),
          type: types.literal('AlignmentsDisplay'),
          configuration: ConfigurationReference(configSchema),
          height: 250,
        })
        .volatile(() => ({
          ReactComponent: AlignmentsDisplayComponent,
        })),
    )
    .views(self => ({
      get pileupDisplayConfig() {
        return {
          ...getConf(self),
          type: 'PileupDisplay',
          name: `${getConf(self, 'name')} pileup`,
          trackId: `${self.configuration.trackId}_pileup`,
        }
      },

      get layoutFeatures() {
        return self.PileupDisplay.layoutFeatures
      },

      get snpCoverageDisplayConfig() {
        return {
          ...getConf(self),
          type: 'SNPCoverageDisplay',
          name: `${getConf(self, 'name')} snpcoverage`,
          trackId: `${self.configuration.trackId}_snpcoverage`,
          adapter: {
            type: 'SNPCoverageAdapter',
            subadapter: getConf(self, 'adapter'),
          },
        }
      },
    }))
    .actions(self => ({
      afterAttach() {
        addDisposer(
          self,
          autorun(() => {
            this.setSNPCoverageDisplay(self.snpCoverageDisplayConfig)
            this.setPileupDisplay(self.pileupDisplayConfig)
          }),
        )
      },
      // eslint-disable-next-line @typescript-eslint/no-explicit-any
      setSNPCoverageDisplay(displayConfig: any) {
        self.SNPCoverageDisplay = {
          type: 'SNPCoverageDisplay',
          configuration: displayConfig,
        }
        self.SNPCoverageDisplay.height = 40
      },
      // eslint-disable-next-line @typescript-eslint/no-explicit-any
      setPileupDisplay(displayConfig: any) {
        self.PileupDisplay = {
          type: 'PileupDisplay',
          configuration: displayConfig,
        }
      },
    }))
}
