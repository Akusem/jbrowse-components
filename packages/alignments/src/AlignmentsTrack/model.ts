import { ConfigurationReference } from '@gmod/jbrowse-core/configuration'
import { BaseTrack } from '@gmod/jbrowse-plugin-linear-genome-view'
import { types } from 'mobx-state-tree'

// eslint-disable-next-line @typescript-eslint/no-explicit-any
export default (pluginManager: any, configSchema: any) => {
  return types.compose(
    'AlignmentsTrack',
    BaseTrack,
    types.model({
      type: types.literal('AlignmentsTrack'),
      configuration: ConfigurationReference(configSchema),
    }),
  )
}
