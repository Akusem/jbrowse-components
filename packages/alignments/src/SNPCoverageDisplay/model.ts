import { types } from 'mobx-state-tree'
import wiggleStateModelFactory from '@gmod/jbrowse-plugin-wiggle/src/WiggleTrack/model'

// using a map because it preserves order
const rendererTypes = new Map([['snpcoverage', 'SNPCoverageRenderer']])

// eslint-disable-next-line @typescript-eslint/no-explicit-any
const stateModelFactory = (configSchema: any) =>
  types.compose(
    'SNPCoverageDisplay',
    wiggleStateModelFactory(configSchema),
    types.model({ type: types.literal('SNPCoverageDisplay') }).views(() => ({
      get rendererTypeName() {
        return rendererTypes.get('snpcoverage')
      },

      get needsScalebar() {
        return true
      },
    })),
  )

export type SNPCoverageDisplayModel = ReturnType<typeof stateModelFactory>

export default stateModelFactory
