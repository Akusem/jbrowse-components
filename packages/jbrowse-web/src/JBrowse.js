import { readConfObject } from '@gmod/jbrowse-core/configuration'
import '@gmod/jbrowse-core/fonts/material-icons.css'
import { App, theme, FatalErrorDialog } from '@gmod/jbrowse-core/ui'
import {
  useDebounce,
  inDevelopment,
  mergeConfigs,
} from '@gmod/jbrowse-core/util'
import ErrorBoundary from 'react-error-boundary'

import { openLocation } from '@gmod/jbrowse-core/util/io'

// material-ui
import CircularProgress from '@material-ui/core/CircularProgress'
import CssBaseline from '@material-ui/core/CssBaseline'
import { ThemeProvider } from '@material-ui/styles'

import { observer } from 'mobx-react'
import { onSnapshot } from 'mobx-state-tree'
import { UndoManager } from 'mst-middlewares'
import React, { useEffect, useState } from 'react'
import 'typeface-roboto'

import JBrowseRootModel from './rootModel'

async function parseConfig(configLoc) {
  const config = JSON.parse(await openLocation(configLoc).readFile('utf8'))
  if (inDevelopment) {
    config.datasets = mergeConfigs(
      config,
      JSON.parse(
        await openLocation({ uri: 'test_data/config_in_dev.json' }).readFile(
          'utf8',
        ),
      ),
    )
  }
  return config
}

function useJBrowseWeb(config, initialState) {
  const [loaded, setLoaded] = useState(false)
  const [rootModel, setRootModel] = useState(initialState || {})

  const [sessionSnapshot, setSessionSnapshot] = useState()
  const debouncedSessionSnapshot = useDebounce(sessionSnapshot, 400)
  const [configSnapshot, setConfigSnapshot] = useState(config || {})

  const { session, jbrowse } = rootModel || {}
  const useLocalStorage = jbrowse
    ? readConfObject(jbrowse.configuration, 'useLocalStorage')
    : false

  // This updates savedSession list on the rootModel
  useEffect(() => {
    if (rootModel && rootModel.session && debouncedSessionSnapshot) {
      rootModel.jbrowse.updateSavedSession(debouncedSessionSnapshot)
    }
  }, [debouncedSessionSnapshot, rootModel])

  useEffect(() => {
    try {
      setRootModel(JBrowseRootModel.create({ jbrowse: configSnapshot }))
    } catch (error) {
      // if it failed to load, it's probably a problem with the saved sessions,
      // so just delete them and try again
      setRootModel(
        JBrowseRootModel.create({
          jbrowse: { ...configSnapshot, savedSessions: [] },
        }),
      )
    }
  }, [configSnapshot])

  // This loads a config from localStorage or a configSnapshot or a config.json file
  useEffect(() => {
    ;(async () => {
      try {
        if (initialState) {
          setRootModel(initialState)
        } else {
          const localStorageConfig =
            useLocalStorage && localStorage.getItem('jbrowse-web-data')

          if (localStorageConfig) {
            setConfigSnapshot(JSON.parse(localStorageConfig))
          }
          if (configSnapshot.uri || configSnapshot.localPath) {
            setConfigSnapshot(await parseConfig(config))
          }
        }
      } catch (e) {
        setLoaded(() => {
          // throw to error boundary
          throw e
        })
      }
    })()
  }, [
    config,
    configSnapshot.localPath,
    configSnapshot.uri,
    initialState,
    useLocalStorage,
  ])

  // finalize rootModel and setLoaded
  useEffect(() => {
    if (rootModel && rootModel.jbrowse) {
      const localStorageSession = localStorage.getItem('jbrowse-web-session')
      if (localStorageSession) {
        rootModel.setSession(JSON.parse(localStorageSession))
      }
      if (!rootModel.session) {
        if (rootModel.jbrowse && rootModel.jbrowse.savedSessions.length) {
          const { name } = rootModel.jbrowse.savedSessions[0]
          rootModel.activateSession(name)
        } else {
          rootModel.setDefaultSession()
        }
      }

      rootModel.setHistory(
        UndoManager.create({}, { targetStore: rootModel.session }),
      )
      setLoaded(true)
    }
  }, [rootModel])

  // make some things available globally for testing
  // e.g. window.MODEL.views[0] in devtools
  useEffect(() => {
    if (loaded) {
      window.MODEL = rootModel.session
      window.ROOTMODEL = rootModel
    }
  }, [loaded, rootModel, rootModel.session])

  // set session in localstorage
  useEffect(
    () =>
      useLocalStorage && loaded
        ? onSnapshot(rootModel.session, snapshot => {
            localStorage.setItem(
              'jbrowse-web-session',
              JSON.stringify(snapshot),
            )
          })
        : () => {},
    [loaded, rootModel, useLocalStorage],
  )

  // save jbrowse-web data in localstorage when the session is snapshotted
  useEffect(
    () =>
      useLocalStorage && loaded
        ? onSnapshot(rootModel.jbrowse, snapshot => {
            localStorage.setItem('jbrowse-web-data', JSON.stringify(snapshot))
          })
        : () => {},
    [loaded, rootModel, useLocalStorage],
  )

  // listen to the session for snapshots
  useEffect(
    () =>
      session
        ? onSnapshot(session, snapshot => {
            setSessionSnapshot(snapshot)
          })
        : () => {},
    [session],
  )

  return [loaded, rootModel]
}

const JBrowse = observer(({ config, initialState }) => {
  const [loaded, root] = useJBrowseWeb(config, initialState)
  const debouncedLoaded = useDebounce(loaded, 400)
  // Use a debounce loaded here to let the circle spinner give a tiny more turn
  // which looks better
  return !debouncedLoaded ? (
    <CircularProgress
      style={{
        position: 'fixed',
        top: '50%',
        left: '50%',
        marginTop: -25,
        marginLeft: -25,
      }}
      size={50}
    />
  ) : (
    <App session={root.session} />
  )
})

export default props => {
  return (
    <ThemeProvider theme={theme}>
      <CssBaseline />
      <ErrorBoundary FallbackComponent={FatalErrorDialog}>
        <JBrowse {...props} />
      </ErrorBoundary>
    </ThemeProvider>
  )
}
