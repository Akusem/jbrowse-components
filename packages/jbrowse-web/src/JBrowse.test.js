import {
  cleanup,
  fireEvent,
  render,
  waitForElement,
} from 'react-testing-library'
import React from 'react'

import fetchMock from 'fetch-mock'
import { LocalFile } from 'generic-filehandle'
import rangeParser from 'range-parser'
import { renderRegion, freeResources } from './workers/rpcMethods'

import model, { createTestSession } from './jbrowseModel'
import JBrowse from './JBrowse'
import config from '../test_data/config_integration_test.json'

fetchMock.config.sendAsJson = false

window.requestIdleCallback = cb => cb()
window.cancelIdleCallback = () => {}

const getFile = url => new LocalFile(require.resolve(`../${url}`))
// fakes server responses from local file object with fetchMock
const readBuffer = async (url, args) => {
  let file
  try {
    file = getFile(url)
  } catch (e) {
    return { status: 404 }
  }
  const maxRangeRequest = 1000000 // kind of arbitrary, part of the rangeParser
  if (args.headers.range) {
    const range = rangeParser(maxRangeRequest, args.headers.range)
    const { start, end } = range[0]
    const len = end - start
    const buf = Buffer.alloc(len)
    const { bytesRead } = await file.read(buf, 0, len, start)
    const stat = await file.stat()
    return {
      status: 206,
      body: buf.slice(0, bytesRead),
      headers: { 'Content-Range': `${start}-${end}/${stat.size}` },
    }
  }
  const body = await file.readFile()
  return { status: 200, body }
}

afterEach(cleanup)

fetchMock.mock('*', readBuffer)

// this is just a little hack to silence a warning that we'll get until react
// fixes this: https://github.com/facebook/react/pull/14853
const originalError = console.error
beforeAll(() => {
  console.error = (...args) => {
    if (/Warning.*not wrapped in act/.test(args[0])) {
      return
    }
    originalError.call(console, ...args)
  }
})

afterAll(() => {
  console.error = originalError
})

describe('valid file tests', () => {
  let jbrowseState
  beforeEach(() => {
    jbrowseState = model.create()
    jbrowseState.configuration.rpc.defaultDriver.set('MainThreadRpcDriver')
    jbrowseState.addSession(config)
  })

  it('access about menu', async () => {
    const { getByText } = render(<JBrowse state={jbrowseState} />)
    await waitForElement(() => getByText('JBrowse'))
    fireEvent.click(getByText('Help'))
    fireEvent.click(getByText('About'))

    const dlg = await waitForElement(() => getByText(/About JBrowse/))
    expect(dlg).toBeTruthy()
  })

  it('click and drag to move sideways', async () => {
    const { getByTestId, getByText } = render(<JBrowse state={jbrowseState} />)
    await waitForElement(() => getByText('ctgA'))
    fireEvent.click(
      await waitForElement(() => getByTestId('volvox_alignments')),
    )
    const start = window.MODEL.views[0].offsetPx
    const track = await waitForElement(() =>
      getByTestId('track-volvox_alignments'),
    )
    fireEvent.mouseDown(track, { clientX: 250, clientY: 20 })
    fireEvent.mouseMove(track, { clientX: 100, clientY: 20 })
    fireEvent.mouseUp(track, { clientX: 100, clientY: 20 })
    const end = window.MODEL.views[0].offsetPx
    expect(end - start).toEqual(150)
  })

  it('opens track selector', async () => {
    const { getByTestId } = render(<JBrowse state={jbrowseState} />)

    await waitForElement(() => getByTestId('volvox_alignments'))
    expect(window.MODEL.views[0].tracks.length).toBe(0)
    fireEvent.click(
      await waitForElement(() => getByTestId('volvox_alignments')),
    )
    expect(window.MODEL.views[0].tracks.length).toBe(1)
  })

  it('opens reference sequence track and expects zoom in message', async () => {
    const { getByTestId, getByText } = render(<JBrowse state={jbrowseState} />)
    fireEvent.click(await waitForElement(() => getByTestId('volvox_refseq')))
    window.MODEL.views[0].setNewView(20, 0)
    await waitForElement(() => getByTestId('track-volvox_refseq'))
    expect(getByText('Zoom in to see sequence')).toBeTruthy()
  })
})

describe('some error state', () => {
  let jbrowseState
  beforeEach(() => {
    jbrowseState = model.create()
    jbrowseState.configuration.rpc.defaultDriver.set('MainThreadRpcDriver')
    jbrowseState.addSession(config)
  })

  it('test that track with 404 file displays error', async () => {
    const { getByTestId, getByText } = render(<JBrowse state={jbrowseState} />)
    fireEvent.click(
      await waitForElement(() => getByTestId('volvox_alignments_nonexist')),
    )
    expect(
      await waitForElement(() =>
        getByText(
          'HTTP 404 fetching /test_data/volvox-sorted.bam.bai.nonexist',
        ),
      ),
    ).toBeTruthy()
  })
  it('test that bam with contigA instead of ctgA displays', async () => {
    const { getByTestId, getByText } = render(<JBrowse configs={[config]} />)
    fireEvent.click(
      await waitForElement(() => getByTestId('volvox_bam_altname')),
    )
    expect(
      await waitForElement(() => getByText('ctgA_110_638_0:0:0_3:0:0_15b')),
    ).toBeTruthy()
  })
})

describe.only('variant', () => {
  it('click on a vcf feature', async () => {
    const { getByTestId: byId, getByText } = render(
      <JBrowse configs={[config]} />,
    )
    await waitForElement(() => getByText('JBrowse'))
    window.MODEL.views[0].setNewView(0.05, 5000)
    fireEvent.click(await waitForElement(() => byId('volvox_filtered_vcf')))
    fireEvent.click(await waitForElement(() => byId('vcf-2560')))
    expect(await waitForElement(() => getByText('ctgA:277..277'))).toBeTruthy()
  })
})

describe('bigwig', () => {
  it('open a bigwig track', async () => {
    const { getByTestId: byId, getByText } = render(
      <JBrowse configs={[config]} />,
    )
    await waitForElement(() => getByText('JBrowse'))
    window.MODEL.views[0].setNewView(0.05, 5000)
    fireEvent.click(await waitForElement(() => byId('volvox_microarray')))
    await waitForElement(() => byId('prerendered_canvas'))
  })
  it('open a bigwig line track', async () => {
    const { getByTestId: byId, getByText } = render(
      <JBrowse configs={[config]} />,
    )
    await waitForElement(() => getByText('JBrowse'))
    window.MODEL.views[0].setNewView(0.05, 5000)
    fireEvent.click(await waitForElement(() => byId('volvox_microarray_line')))
    await waitForElement(() => byId('prerendered_canvas'))
  })
  it('open a bigwig density track', async () => {
    const { getByTestId: byId, getByText } = render(
      <JBrowse configs={[config]} />,
    )
    await waitForElement(() => getByText('JBrowse'))
    window.MODEL.views[0].setNewView(0.05, 5000)
    fireEvent.click(
      await waitForElement(() => byId('volvox_microarray_density')),
    )
    await waitForElement(() => byId('prerendered_canvas'))
  })
})

const baseprops = {
  region: { refName: 'ctgA', start: 0, end: 800 },
  sessionId: 'knickers the cow',
  adapterType: 'BamAdapter',
  adapterConfig: {
    configId: '7Hc9NkuD4x',
    type: 'BamAdapter',
    bamLocation: {
      localPath: require.resolve('../public/test_data/volvox-sorted.bam'),
    },
    index: {
      configId: 'sGW8va26pr',
      location: {
        localPath: require.resolve('../public/test_data/volvox-sorted.bam.bai'),
      },
    },
  },
  rootConfig: {},
  renderProps: { bpPerPx: 1 },
}

describe('render tests', () => {
  let pluginManager
  beforeEach(() => {
    ;({ pluginManager } = createTestSession())
  })

  it('can render a single region with Pileup + BamAdapter', async () => {
    const testprops = { ...baseprops, rendererType: 'PileupRenderer' }

    const result = await renderRegion(pluginManager, testprops)
    expect(new Set(Object.keys(result))).toEqual(
      new Set(['features', 'html', 'layout', 'height', 'width', 'imageData']),
    )
    expect(result.features.length).toBe(93)
    expect(result.html).toMatchSnapshot()
    expect(result.layout).toMatchSnapshot()
    expect(result.imageData.width).toBe(800)
    expect(result.imageData.height).toBe(result.layout.totalHeight)
    expect(result.width).toBe(800)
    expect(result.height).toBe(result.layout.totalHeight)

    expect(
      freeResources(pluginManager, {
        sessionId: 'knickers the cow',
      }),
    ).toBe(2)

    expect(
      freeResources(pluginManager, {
        sessionId: 'fozzy bear',
      }),
    ).toBe(0)
  })

  it('can render a single region with SvgFeatures + BamAdapter', async () => {
    const testprops = {
      ...baseprops,
      rendererType: 'SvgFeatureRenderer',
      region: { refName: 'ctgA', start: 0, end: 300 },
    }

    const result = await renderRegion(pluginManager, testprops)
    expect(new Set(Object.keys(result))).toEqual(
      new Set(['html', 'features', 'layout']),
    )
    expect(result.features.length).toBe(25)
    expect(result.html).toMatchSnapshot()
    expect(result.layout).toMatchSnapshot()

    expect(
      freeResources(pluginManager, {
        sessionId: 'knickers the cow',
      }),
    ).toBe(2)

    expect(
      freeResources(pluginManager, {
        sessionId: 'fozzy bear',
      }),
    ).toBe(0)
  })

  it('throws if no session ID', async () => {
    const testprops = {
      ...baseprops,
      rendererType: 'PileupRenderer',
      sessionId: undefined,
    }

    await expect(renderRegion(pluginManager, testprops)).rejects.toThrow(
      /must pass a unique session id/,
    )
  })

  it('throws on unrecoginze worker', async () => {
    const testprops = {
      ...baseprops,
      rendererType: 'NonexistentRenderer',
    }

    await expect(renderRegion(pluginManager, testprops)).rejects.toThrow(
      /renderer "NonexistentRenderer" not found/,
    )
  })
})
