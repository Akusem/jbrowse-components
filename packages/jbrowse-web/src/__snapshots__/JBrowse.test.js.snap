// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`render tests can render a single region with Pileup + BamAdapter 2`] = `
Object {
  "rectangles": Object {
    "4521984": Array [
      2,
      0,
      102,
      7,
    ],
    "4522237": Array [
      7,
      9,
      107,
      16,
    ],
    "4522500": Array [
      15,
      18,
      115,
      25,
    ],
    "4522762": Array [
      16,
      27,
      116,
      34,
    ],
    "4523023": Array [
      17,
      36,
      117,
      43,
    ],
    "4523283": Array [
      42,
      45,
      142,
      52,
    ],
    "4523538": Array [
      43,
      54,
      143,
      61,
    ],
    "4523795": Array [
      47,
      63,
      147,
      70,
    ],
    "4524052": Array [
      55,
      72,
      155,
      79,
    ],
    "4524307": Array [
      65,
      81,
      165,
      88,
    ],
    "4524562": Array [
      76,
      90,
      176,
      97,
    ],
    "4524815": Array [
      90,
      99,
      190,
      106,
    ],
    "4525068": Array [
      94,
      108,
      194,
      115,
    ],
    "4525322": Array [
      109,
      0,
      209,
      7,
    ],
    "4525576": Array [
      123,
      9,
      223,
      16,
    ],
    "4525835": Array [
      125,
      18,
      225,
      25,
    ],
    "4526091": Array [
      132,
      27,
      232,
      34,
    ],
    "4526356": Array [
      140,
      36,
      240,
      43,
    ],
    "4526613": Array [
      147,
      45,
      247,
      52,
    ],
    "4526869": Array [
      149,
      54,
      249,
      61,
    ],
    "4527129": Array [
      152,
      63,
      252,
      70,
    ],
    "4527384": Array [
      186,
      72,
      286,
      79,
    ],
    "4527644": Array [
      237,
      0,
      337,
      7,
    ],
    "4527900": Array [
      258,
      9,
      358,
      16,
    ],
    "4528158": Array [
      291,
      18,
      391,
      25,
    ],
    "4528415": Array [
      301,
      27,
      401,
      34,
    ],
    "4528674": Array [
      302,
      36,
      402,
      43,
    ],
    "4528935": Array [
      311,
      45,
      411,
      52,
    ],
    "4529189": Array [
      321,
      54,
      421,
      61,
    ],
    "4529446": Array [
      325,
      63,
      425,
      70,
    ],
    "4529704": Array [
      327,
      72,
      427,
      79,
    ],
    "4529960": Array [
      338,
      0,
      438,
      7,
    ],
    "4530218": Array [
      354,
      81,
      454,
      88,
    ],
    "4530479": Array [
      390,
      9,
      490,
      16,
    ],
    "4530736": Array [
      399,
      18,
      499,
      25,
    ],
    "4530996": Array [
      413,
      27,
      513,
      34,
    ],
    "4531253": Array [
      413,
      36,
      513,
      43,
    ],
    "4531511": Array [
      422,
      45,
      522,
      52,
    ],
    "4531768": Array [
      431,
      54,
      531,
      61,
    ],
    "4532023": Array [
      432,
      63,
      532,
      70,
    ],
    "4532283": Array [
      453,
      0,
      553,
      7,
    ],
    "4532542": Array [
      461,
      72,
      561,
      79,
    ],
    "4532795": Array [
      467,
      81,
      567,
      88,
    ],
    "4533053": Array [
      481,
      90,
      581,
      97,
    ],
    "4533313": Array [
      485,
      99,
      585,
      106,
    ],
    "4533575": Array [
      487,
      108,
      587,
      115,
    ],
    "4533834": Array [
      495,
      9,
      595,
      16,
    ],
    "4534100": Array [
      509,
      18,
      609,
      25,
    ],
    "4534362": Array [
      513,
      117,
      613,
      124,
    ],
    "4534621": Array [
      540,
      27,
      640,
      34,
    ],
    "4534877": Array [
      545,
      36,
      645,
      43,
    ],
    "4535134": Array [
      556,
      0,
      656,
      7,
    ],
    "4535394": Array [
      562,
      45,
      662,
      52,
    ],
    "4535651": Array [
      564,
      54,
      664,
      61,
    ],
    "4535906": Array [
      564,
      63,
      664,
      70,
    ],
    "4536165": Array [
      564,
      72,
      664,
      79,
    ],
    "4536423": Array [
      566,
      126,
      666,
      133,
    ],
    "4536689": Array [
      570,
      81,
      670,
      88,
    ],
    "4536952": Array [
      574,
      135,
      674,
      142,
    ],
    "4537214": Array [
      574,
      144,
      674,
      151,
    ],
    "4537471": Array [
      577,
      153,
      677,
      160,
    ],
    "4537728": Array [
      579,
      162,
      679,
      169,
    ],
    "4537984": Array [
      582,
      90,
      682,
      97,
    ],
    "4538240": Array [
      593,
      99,
      693,
      106,
    ],
    "4538496": Array [
      595,
      108,
      695,
      115,
    ],
    "4538753": Array [
      611,
      9,
      711,
      16,
    ],
    "4539018": Array [
      611,
      18,
      711,
      25,
    ],
    "4539274": Array [
      612,
      171,
      712,
      178,
    ],
    "4539531": Array [
      618,
      117,
      718,
      124,
    ],
    "4539793": Array [
      625,
      180,
      725,
      187,
    ],
    "4540054": Array [
      626,
      189,
      726,
      196,
    ],
    "4540318": Array [
      630,
      198,
      730,
      205,
    ],
    "4540572": Array [
      630,
      207,
      730,
      214,
    ],
    "4540835": Array [
      652,
      27,
      752,
      34,
    ],
    "4541095": Array [
      660,
      0,
      760,
      7,
    ],
    "4541356": Array [
      661,
      36,
      761,
      43,
    ],
    "4541614": Array [
      664,
      45,
      764,
      52,
    ],
    "4541876": Array [
      679,
      54,
      779,
      61,
    ],
    "4542131": Array [
      685,
      63,
      785,
      70,
    ],
    "4542390": Array [
      690,
      72,
      790,
      79,
    ],
    "4542645": Array [
      713,
      9,
      813,
      16,
    ],
    "4542907": Array [
      715,
      18,
      815,
      25,
    ],
    "4543166": Array [
      720,
      81,
      820,
      88,
    ],
    "4543425": Array [
      725,
      90,
      825,
      97,
    ],
    "4543681": Array [
      727,
      99,
      827,
      106,
    ],
    "4543943": Array [
      729,
      108,
      829,
      115,
    ],
    "4544200": Array [
      730,
      117,
      830,
      124,
    ],
    "4544461": Array [
      747,
      126,
      847,
      133,
    ],
    "4544716": Array [
      764,
      0,
      864,
      7,
    ],
    "4544974": Array [
      766,
      27,
      866,
      34,
    ],
    "4545233": Array [
      775,
      36,
      875,
      43,
    ],
    "4545496": Array [
      786,
      45,
      886,
      52,
    ],
    "4545752": Array [
      797,
      54,
      897,
      61,
    ],
  },
  "totalHeight": 216,
}
`;

exports[`render tests can render a single region with SvgFeatures + BamAdapter 2`] = `
Object {
  "rectangles": Object {
    "4521984": Array [
      2,
      0,
      195,
      25,
    ],
    "4522237": Array [
      7,
      27,
      200,
      52,
    ],
    "4522500": Array [
      15,
      54,
      215,
      79,
    ],
    "4522762": Array [
      16,
      81,
      209,
      106,
    ],
    "4523023": Array [
      17,
      108,
      217,
      133,
    ],
    "4523283": Array [
      42,
      135,
      235,
      160,
    ],
    "4523538": Array [
      43,
      162,
      243,
      187,
    ],
    "4523795": Array [
      47,
      189,
      247,
      214,
    ],
    "4524052": Array [
      55,
      216,
      255,
      241,
    ],
    "4524307": Array [
      65,
      243,
      258,
      268,
    ],
    "4524562": Array [
      76,
      270,
      269,
      295,
    ],
    "4524815": Array [
      90,
      297,
      283,
      322,
    ],
    "4525068": Array [
      94,
      324,
      294,
      349,
    ],
    "4525322": Array [
      109,
      351,
      309,
      376,
    ],
    "4525576": Array [
      123,
      378,
      330,
      403,
    ],
    "4525835": Array [
      125,
      405,
      325,
      430,
    ],
    "4526091": Array [
      132,
      432,
      339,
      457,
    ],
    "4526356": Array [
      140,
      459,
      347,
      484,
    ],
    "4526613": Array [
      147,
      486,
      347,
      511,
    ],
    "4526869": Array [
      149,
      513,
      342,
      538,
    ],
    "4527129": Array [
      152,
      540,
      359,
      565,
    ],
    "4527384": Array [
      186,
      567,
      386,
      592,
    ],
    "4527644": Array [
      237,
      0,
      437,
      25,
    ],
    "4527900": Array [
      258,
      27,
      458,
      52,
    ],
    "4528158": Array [
      291,
      54,
      498,
      79,
    ],
    "4528415": Array [
      301,
      81,
      508,
      106,
    ],
    "4528674": Array [
      302,
      108,
      502,
      133,
    ],
    "4528935": Array [
      311,
      135,
      511,
      160,
    ],
    "4529189": Array [
      321,
      162,
      528,
      187,
    ],
    "4529446": Array [
      325,
      189,
      525,
      214,
    ],
    "4529704": Array [
      327,
      216,
      527,
      241,
    ],
    "4529960": Array [
      338,
      243,
      538,
      268,
    ],
    "4530218": Array [
      354,
      270,
      554,
      295,
    ],
    "4530479": Array [
      390,
      297,
      597,
      322,
    ],
    "4530736": Array [
      399,
      324,
      599,
      349,
    ],
    "4530996": Array [
      413,
      351,
      620,
      376,
    ],
    "4531253": Array [
      413,
      378,
      606,
      403,
    ],
    "4531511": Array [
      422,
      405,
      629,
      430,
    ],
    "4531768": Array [
      431,
      432,
      624,
      457,
    ],
    "4532023": Array [
      432,
      459,
      625,
      484,
    ],
    "4532283": Array [
      453,
      0,
      653,
      25,
    ],
    "4532542": Array [
      461,
      27,
      647,
      52,
    ],
    "4532795": Array [
      467,
      486,
      667,
      511,
    ],
    "4533053": Array [
      481,
      513,
      674,
      538,
    ],
    "4533313": Array [
      485,
      540,
      685,
      565,
    ],
    "4533575": Array [
      487,
      567,
      687,
      592,
    ],
    "4533834": Array [
      495,
      594,
      702,
      619,
    ],
    "4534100": Array [
      509,
      54,
      716,
      79,
    ],
    "4534362": Array [
      513,
      81,
      713,
      106,
    ],
    "4534621": Array [
      540,
      108,
      747,
      133,
    ],
    "4534877": Array [
      545,
      135,
      752,
      160,
    ],
    "4535134": Array [
      556,
      162,
      763,
      187,
    ],
    "4535394": Array [
      562,
      189,
      769,
      214,
    ],
    "4535651": Array [
      564,
      216,
      757,
      241,
    ],
    "4535906": Array [
      564,
      243,
      771,
      268,
    ],
    "4536165": Array [
      564,
      270,
      779,
      295,
    ],
    "4536423": Array [
      566,
      621,
      766,
      646,
    ],
    "4536689": Array [
      570,
      648,
      777,
      673,
    ],
    "4536952": Array [
      574,
      675,
      781,
      700,
    ],
    "4537214": Array [
      574,
      702,
      789,
      727,
    ],
    "4537471": Array [
      577,
      729,
      784,
      754,
    ],
    "4537728": Array [
      579,
      756,
      779,
      781,
    ],
    "4537984": Array [
      582,
      783,
      797,
      808,
    ],
    "4538240": Array [
      593,
      810,
      793,
      835,
    ],
    "4538496": Array [
      595,
      837,
      802,
      862,
    ],
    "4538753": Array [
      611,
      297,
      826,
      322,
    ],
    "4539018": Array [
      611,
      324,
      826,
      349,
    ],
    "4539274": Array [
      612,
      378,
      819,
      403,
    ],
    "4539531": Array [
      618,
      864,
      818,
      889,
    ],
    "4539793": Array [
      625,
      351,
      825,
      376,
    ],
    "4540054": Array [
      626,
      432,
      841,
      457,
    ],
    "4540318": Array [
      630,
      405,
      830,
      430,
    ],
    "4540572": Array [
      630,
      459,
      830,
      484,
    ],
    "4540835": Array [
      652,
      27,
      859,
      52,
    ],
    "4541095": Array [
      660,
      0,
      875,
      25,
    ],
    "4541356": Array [
      661,
      891,
      876,
      916,
    ],
    "4541614": Array [
      664,
      918,
      871,
      943,
    ],
    "4541876": Array [
      679,
      486,
      886,
      511,
    ],
    "4542131": Array [
      685,
      513,
      885,
      538,
    ],
    "4542390": Array [
      690,
      540,
      890,
      565,
    ],
    "4542645": Array [
      713,
      567,
      920,
      592,
    ],
    "4542907": Array [
      715,
      81,
      922,
      106,
    ],
    "4543166": Array [
      720,
      54,
      927,
      79,
    ],
    "4543425": Array [
      725,
      594,
      940,
      619,
    ],
    "4543681": Array [
      727,
      945,
      934,
      970,
    ],
    "4543943": Array [
      729,
      972,
      936,
      997,
    ],
    "4544200": Array [
      730,
      999,
      930,
      1024,
    ],
    "4544461": Array [
      747,
      1026,
      954,
      1051,
    ],
    "4544716": Array [
      764,
      108,
      964,
      133,
    ],
    "4544974": Array [
      766,
      135,
      973,
      160,
    ],
    "4545233": Array [
      775,
      162,
      990,
      187,
    ],
    "4545496": Array [
      786,
      189,
      1001,
      214,
    ],
    "4545752": Array [
      797,
      216,
      1004,
      241,
    ],
  },
  "totalHeight": 1053,
}
`;
